### ISSUE TITLE
Plural DbContext names, singular tables?

### ISSUE THREAD

**Issue by atrauzzi** [efcore/EFCore.NamingConventions#15]
As the title says, I think one convention that might be worth having as a toggle is to inflect singular table names from plural `DbContext` properties.

🙂

**Comment by roji** [comment]
@bricelam I know we occasionally mention pluralization but I don't have a good picture of this... Does pluralization/singularization belong in a naming conventions plugin like this one?

**Comment by bricelam** [comment]
We don’t do it at runtime in the product because fixing any pluralization bugs would be a breaking change. We gave plans to do it at design time for reverse engineering. IMHO, it’s totally fine in an extension like this where the bar for breaking changes is a lot lower.

But, just removing the convention that uses the DbSet name as the table name would be enough for this issue, right?

**Comment by atrauzzi** [comment]
I think the idea is that a convention is still desired, just that it inflects for table names as .net prescribes plural `DbSet`s and some DBs use singular, snake cased equivalents.

**Comment by roji** [comment]
I agree singularization should be OK in this extension, anyone know of a good singularizer?

**Comment by bricelam** [comment]
The Roslyn team uses https://github.com/Humanizr/Humanizer for name reccomendations

**Comment by bricelam** [comment]
But wouldn't doing https://github.com/dotnet/efcore/issues/11160 light up with this extension?

**Comment by nerdpad** [comment]
I had to use `Humanizer` in my project to achieve this. It will be great if it was part of this extension.

**Comment by VitalyV** [comment]
Do we have any progress?

**Comment by roji** [comment]
Not at the moment, there's too many other things going on. I can try to review a PR though.

**Comment by MacShack** [comment]
Hold on a moment. Am I the only one who just wants the name of the tables to reflect the name of the type rather than that of some singularized form of a plural name which most of the time is already a pluralized form of the type?

So you have a entity called `Person` (which most likely is a wrong name for the type anyway. But ok). Then people spend time debating really hard that it should be `People` instead of `Persons`. But have at least agreed with each other that all table names should be in singular from. So they are back to square one.

Shouldn't we just add an option to use the type name as naming: `UseTypeNameForTableNamingConvention` and leave out all the complexity of some `Humanizer`?

**Comment by roji** [comment]
@MacShack I think you have things confused - this extension (EFCore.NamingConventions) just changes the casing of identifiers (e.g. PascalCase to snake-case), it does not do any plural->singular transformations (or vice versa). This specific issue is asking for this extension to do singularization (but it does not currently do that).

As for EF itself, as @bricelam noted [above](https://github.com/efcore/EFCore.NamingConventions/issues/15#issuecomment-576391788), the pluralization/singularization is done at runtime.

So I think things already work as you want?

**Comment by jeremycook** [comment]
You can easily remove the pluralization convention in EF Core 6 onward with the following.

```csharp
// In your DbContext
protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
{
    configurationBuilder.Conventions.Remove(typeof(TableNameFromDbSetConvention));
}
```

* Official docs: https://learn.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration
* An EF Core community standup video that covers this capability in great depth: https://www.youtube.com/live/6apfe1L1FhY

**Comment by atrauzzi** [comment]
Will that make the table names singular, or just result in an error as a result of them not being explicitly specified per-model?

**Comment by jeremycook** [comment]
@atrauzzi EF Core doesn't need the "TableNameFromDbSetConvention" convention to name tables. If that convention is removed it will use the name of the class if the [Table] attribute has not been applied to it. I use the technique I mentioned in projects where I have a mixture of EF Core 6+ and Dapper for example.

**Comment by LeTranAnhVu** [comment]
@jeremycook Your suggestion did work in my case. Thank you!

**Comment by buden** [comment]
@jeremycook is that possible to use your solution in EF6? The ModelConfigurationBuilder.Conventions property is not available there.

**Comment by jeremycook** [comment]
@buden It looks like I misspoke about it being available in EF Core 6. Looking at [that version of the docs](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.modelconfigurationbuilder?view=efcore-6.0), I also don't see the `ModelConfigurationBuilder.Conventions` property. It is there in the [EF Core 7 version of the docs](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.modelconfigurationbuilder?view=efcore-7.0).

**Comment by kalatchev** [comment]
@atrauzzi 

IMHO good practice for DB design is to name tables in plural form, e.g. if your entity is called `Sale`, the table have to be `sales` (in PG). The reason is that **this table contains many sales**. As far as I remember, that comes from many years, back to the days of first publications about [Active Record pattern](https://en.wikipedia.org/wiki/Active_record_pattern). EF is more or less the implementation of this pattern, i.e. should follow its logic.

**Comment by markushaslinger** [comment]
> You can easily remove the pluralization convention in EF Core 7 onward with the following.
> 
> ```cs
> // In your DbContext
> protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
> {
>     configurationBuilder.Conventions.Remove(typeof(TableNameFromDbSetConvention));
> }
> ```

Note: this breaks when using TPC, the concrete tables generated are still pluralized leading to inconsistency 

IMHO a table is the entity type (singular) while the set of tuples is the entity set (plural). 
It even shows in how we define `DbSet<Car> Cars`: a set of type `Car` called `Cars`.

I really hope we will finally get a simple way to _reliably_ override this default.

**Comment by dkezri** [comment]
Thanks @haslingerm 

i had same problem when upgraded to dotnet 9.0  ,  and this helped to fix my plural naming av tables.

**Comment by vlademirjunior** [comment]
@jeremycook the GOAT! thanks bro!
